### PR TITLE
Add CodeableReference keyword

### PR DESCRIPTION
**Pull request by jafeltra**
This PR adds a `CodeableReference` keyword, which can be used in Type Rules to constrain a type to a CodeableReference or a specific set of CodeableReference resource/profiles and Add Element Rules to add a CodeableReference element.

As of right now, the `CodeableReference` keyword cannot be used in AssignmentRules. To assign a portion of a `CodeableReference` element, you can assign the `concept` and/or `reference` portion directly, as you do now in the current version of SUSHI. If it turns out we do want to add AssignmentRule use for the `CodeableReference` PR, I'll either close this PR or mark it as draft while making the updates.

### CODE REVIEW

File: test/fhirtypes/ElementDefinition.constrainType.test.ts
Comment (mint-thompson) [inline test/fhirtypes/ElementDefinition.constrainType.test.ts:1]: It could be useful to add some tests that constrain an element to a combination of Reference/CodeableReference and non-reference types.

**Review by mint-thompson — CHANGES_REQUESTED** [review]
This looks really good, and should definitely help with using CodeableReference types. I have a few small suggestions.

File: src/fhirtypes/ElementDefinition.ts [hunk 1]
```diff
@@ -1171,7 +1199,7 @@ export class ElementDefinition {
         // one of the targetProfiles.  If the targetProfile property is null, that means any
         // reference is allowed.
         // When 'Reference' keyword is used, prefer to match on the 'Reference' type over the
-        // 'CodeableReference' type if they both exist on the element.
+        // 'CodeableReference' type if they both exist on the element, but issue a warning.
```
[the remark this answers is not in the document]
Comment (jafeltra) [reply src/fhirtypes/ElementDefinition.ts:1202]: opps, yes I moved the warning and forgot about the comment!

File: src/fhirtypes/ElementDefinition.ts [hunk 2]
```diff
@@ -947,6 +960,21 @@ export class ElementDefinition {
     const targetType = this.getTargetType(target, fisher);
     const targetTypes: ElementDefinitionType[] = targetType ? [targetType] : this.type;
 
+    // If the target type is a CodeableReference but the rule types were set via the Reference() keyword,
+    // log a warning to use CodeableReference keyword
+    if (
+      targetTypes.some(t => t.code === 'CodeableReference') &&
+      !targetTypes.some(t => t.code === 'Reference') &&
+      rule.types.every(t => t.isReference)
+    ) {
```
[the remark this answers is not in the document]
Comment (jafeltra) [reply src/fhirtypes/ElementDefinition.ts:969]: Ah, that's a good point, and a case I hadn't thought about. If I just change the condition here, the warning message actually constructs the wrong example because it tries to create `CodeableReference(boolean or Patient)`, which is wrong. So I think I'll also just remove the example, especially since the line number is included as well.

File: test/fhirtypes/ElementDefinition.constrainType.test.ts
Comment (jafeltra) [reply test/fhirtypes/ElementDefinition.constrainType.test.ts:1]: Added in [20614c4](https://github.com/FHIR/sushi/pull/1292/commits/20614c424ca05afcf4c971a81cf8eb387c4a11c3)

File: src/fhirtypes/ElementDefinition.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jafeltra) [reply src/fhirtypes/ElementDefinition.ts]: Removed in [20614c4](https://github.com/FHIR/sushi/pull/1292/commits/20614c424ca05afcf4c971a81cf8eb387c4a11c3)

File: src/fhirtypes/ElementDefinition.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jafeltra) [reply src/fhirtypes/ElementDefinition.ts]: I updated the condition and the warning message and included a test in [20614c4](https://github.com/FHIR/sushi/pull/1292/commits/20614c424ca05afcf4c971a81cf8eb387c4a11c3).

**Review by mint-thompson — CHANGES_REQUESTED** [review]
Thank you for the updates! I have a suggestion about the warning message, but everything else looks good.

File: src/fhirtypes/ElementDefinition.ts [hunk 3]
```diff
@@ -947,6 +960,19 @@ export class ElementDefinition {
...
 
+    // If the target type is a CodeableReference but the rule types were set via the Reference() keyword,
+    // log a warning to use CodeableReference keyword
+    if (
+      targetTypes.some(t => t.code === 'CodeableReference') &&
+      !targetTypes.some(t => t.code === 'Reference') &&
+      rule.types.some(t => t.isReference)
+    ) {
+      logger.warn(
+        'The CodeableReference() syntax should be used to constrain references of a CodeableReference',
```
[the remark this answers is not in the document]
Comment (jafeltra) [reply src/fhirtypes/ElementDefinition.ts]: I like it. Added in [6ef48c8](https://github.com/FHIR/sushi/pull/1292/commits/6ef48c816115d3770932475600abd86ecce8b943).

**Comment by cmoesel** [comment]
FYI - I've discussed with Mark Kramer and he agrees with the approach here: `CodeableConcept()` is suitable for constraining but should not be used for assignments.

File: test/fhirtypes/ElementDefinition.constrainType.test.ts [hunk 4]
```diff
@@ -1207,102 +1207,288 @@ describe('ElementDefinition R5', () => {
...
-
-    it('should throw InvalidTypeError when a passed in reference to a type that cannot constrain any existing references to types on a CodeableReference', () => {
-      const addresses = r5CarePlan.elements.find(e => e.id === 'CarePlan.addresses');
-      const onlyRule = new OnlyRule('addresses');
-      onlyRule.types = [{ type: 'Patient', isReference: true }];
-      expect(() => {
-        addresses.constrainType(onlyRule, fisher);
-      }).toThrow(/"Reference\(Patient\).*Reference\(.*Condition\)/);
+    describe('Reference() keyword', () => {
+      it('should allow a CodeableReference to multiple resource types to be constrained to a reference to a subset', () => {
```
Comment (cmoesel) [inline test/fhirtypes/ElementDefinition.constrainType.test.ts:1215]: I'd suggest rewording the tests that use Reference to constrain a CodeableConcept so that the test summary indicates that a warning should be logged. E.g.:
```suggestion
      it('should warn when a CodeableReference to multiple resource types is constrained to a Reference to a subset', () => {
```

File: test/fhirtypes/ElementDefinition.constrainType.test.ts [hunk 5]
```diff
@@ -1207,102 +1207,288 @@ describe('ElementDefinition R5', () => {
...
+          )
+        );
+        expect(loggerSpy.getAllMessages('error')).toHaveLength(0);
+        expect(loggerSpy.getAllMessages('warn')).toHaveLength(1);
+        expect(loggerSpy.getLastMessage('warn')).toMatch(
+          /CodeableReference\(\) keyword should be used/i
+        );
+      });
+
+      it('should allow a CodeableReference to multiple resource types to be constrained to a reference to a single profile', () => {
```
Comment (cmoesel) [inline test/fhirtypes/ElementDefinition.constrainType.test.ts:1281]:
```suggestion
      it('should allow a CodeableReference to multiple resource types to be constrained to a reference to multiple profiles', () => {
```

File: test/fhirtypes/ElementDefinition.constrainType.test.ts [hunk 6]
```diff
@@ -1207,102 +1207,288 @@ describe('ElementDefinition R5', () => {
...
+        expect(performedActivity.type[0]).toEqual(
+          new ElementDefinitionType('CodeableReference').withTargetProfiles(
+            'http://hl7.org/fhir/StructureDefinition/actualgroup'
+          )
+        );
+        expect(loggerSpy.getAllMessages('error')).toHaveLength(0);
+        expect(loggerSpy.getAllMessages('warn')).toHaveLength(0);
+      });
+
+      it('should allow a CodeableReference to multiple resource types to be constrained to a reference to a single profile using CodeableReference keyword', () => {
```
Comment (cmoesel) [inline test/fhirtypes/ElementDefinition.constrainType.test.ts:1417]:
```suggestion
      it('should allow a CodeableReference to multiple resource types to be constrained to a reference to multiple profiles using CodeableReference keyword', () => {
```

File: test/import/FSHImporter.Profile.test.ts [hunk 7]
```diff
@@ -1681,6 +1681,58 @@ describe('FSHImporter', () => {
...
+          { type: 'Organization', isCodeableReference: true },
+          { type: 'CareTeam', isCodeableReference: true }
+        );
+      });
+
+      it('should parse an only rule with a CodeableReference to multiple types with whitespace', () => {
+        const input = leftAlign(`
+        Profile: ObservationProfile
+        Parent: Observation
+        * performer only CodeableReference(   Organization    or  CareTeam)
```
Comment (cmoesel) [inline test/import/FSHImporter.Profile.test.ts:1722]: Just to be sure we're allowing white space in all the places, I'd suggest adding some before the last `)` too:
```suggestion
        * performer only CodeableReference(   Organization    or  CareTeam       )
```

**Review by mint-thompson — APPROVED** [review]

File: test/import/FSHImporter.SDRules.test.ts [hunk 8]
```diff
@@ -1319,6 +1319,58 @@ describe('FSHImporter', () => {
         );
       });
 
+      it('should parse an only rule with a CodeableReference to one type', () => {
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply test/import/FSHImporter.SDRules.test.ts:1322]: If you can confirm that these are exactly the same, I'm fine with removing them. We may have overlapping tests due to previous refactoring... Or in a few cases, I think we have a token test cases in broad test files and then many detailed ones in more specific test files. Maybe that is what's going on here? Either way, duplicate tests aren't really helpful so I'm fine w/ them going away.

That said, if there are lots of duplicate tests in this file in general, then it may be worth making a separate task to clean the whole thing up.

**Review by cmoesel — CHANGES_REQUESTED** [review]
This looks good and works well. I've mainly just made some small suggestions for the tests.

File: test/import/FSHImporter.Profile.test.ts (hunk 7 above)
Comment (jafeltra) [reply test/import/FSHImporter.Profile.test.ts:1722]: Good call!

File: test/import/FSHImporter.SDRules.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jafeltra) [reply test/import/FSHImporter.SDRules.test.ts:1322]: I double checked, and as far as I can tell, they are exactly the same. Neither set of tests sets up anything different in `describe` and/or `beforeAll`/`beforeEach` blocks.

There do seem to be large amounts of duplicates between these two files. The `FSHImporter.SDRules.tests.ts` file does mention at the top that it is designed to cover the overlap between the four different StructureDefinition entities (profiles, extensions, resources, and logicals). I looked at the OnlyRule section, where the tests I added were, as well as a few other sections. There is a large amount of overlap between the tests, but unfortunately there are a few tests that were newly added in both files. For example, the CaretValueRule section has 4 tests shared between the two files, 1 new test added in the FSHImporter.Profile file, and 1 new test added in the FSHImporter.SDRules file.

So that is... unfortunate. It seems like the files might have been refactored when Logicals and Resources were added, so I wonder if something got messed up in the rebasing/merging back then. Who knows.

All that said, I think what makes sense is to add the three new tests I included here into the `FSHImporter.SDRules.test.ts` file and remove them from the `FSHImporter.Profile.test.ts` file since they apply to any StructureDefinition and that will reduce clutter for now. And then it probably is a good idea to clean up these files so there are fewer shared tests between them.

[the remark this answers is not in the document]
Comment (jafeltra) [reply test/import/FSHImporter.SDRules.test.ts:1322]: I checked the InsertRule block and there's a whole bunch of tests in the `FSHImporter.Profile.test.ts` that aren't in the `FSHImporter.SDRules.test.ts` (~12 tests). So I actually don't know for sure which file we should be trying to maintain, but I think the `FSHImporter.SDRules.test.ts` makes more sense as a centralized place to hold the overlapping tests. So I'll still plan to move my tests to be in `FSHImporter.SDRules.test.ts`, but I can be convinced otherwise if either of you disagree on the file to hold them.

File: test/import/FSHImporter.Profile.test.ts (hunk 7 above)
Comment (jafeltra) [reply test/import/FSHImporter.Profile.test.ts]: Updated in [9f88612](https://github.com/FHIR/sushi/pull/1292/commits/9f88612248639230232f7cb9157c6dd3bb37663b), but then promptly removed in favor of the tests in `FSHImporter.SDRules.test.ts`.

File: test/fhirtypes/ElementDefinition.constrainType.test.ts (hunk 6 above)
Comment (jafeltra) [reply test/fhirtypes/ElementDefinition.constrainType.test.ts]: Updated in [9f88612](https://github.com/FHIR/sushi/pull/1292/commits/9f88612248639230232f7cb9157c6dd3bb37663b)

File: test/fhirtypes/ElementDefinition.constrainType.test.ts (hunk 5 above)
Comment (jafeltra) [reply test/fhirtypes/ElementDefinition.constrainType.test.ts]: Updated in [9f88612](https://github.com/FHIR/sushi/pull/1292/commits/9f88612248639230232f7cb9157c6dd3bb37663b)

File: test/fhirtypes/ElementDefinition.constrainType.test.ts (hunk 4 above)
Comment (jafeltra) [reply test/fhirtypes/ElementDefinition.constrainType.test.ts]: Updated in [9f88612](https://github.com/FHIR/sushi/pull/1292/commits/9f88612248639230232f7cb9157c6dd3bb37663b), along with the few other tests that needed this update.

File: test/import/FSHImporter.SDRules.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jafeltra) [reply test/import/FSHImporter.SDRules.test.ts:1322]: Removed the duplicate tests in [7d6c098](https://github.com/FHIR/sushi/pull/1292/commits/7d6c098e0d9d0c1a5a45c4093d60edcf674df3b5), but I can swap the location once you have a chance to process the novels I wrote above.

[the remark this answers is not in the document]
Comment (cmoesel) [reply test/import/FSHImporter.SDRules.test.ts:1322]: I don't know. On one hand, I understand the approach of having tests for logic that applies to any SD being in `FSHImporter.SDRules.test.ts`. On the other hand, when you write the test, you _do_ have to choose a specific kind of SD, right? But when we're talking about importing rules, I guess it shouldn't matter if it is a profile, logical, etc...  So... ???

I think one of the reasons we split into `FSHImporter.Profile.test.ts` and others was to keep `FSHImporter.SDRules.test.ts` from becoming massive. But at the same time, if we put a rule import test _only_ in `FSHImporter.Profile.test.ts`, then it's not easy to know where to look for it and people might thing that type of rule is not tested at all.

All this to say, I think you can justify either way. But I guess maybe if I had to lean, I'd lean toward `FSHImporter.SDRules.test.ts`? Maybe we should refactor into a separate file for every rule type instead. That actually might make more sense, but is way outside the scope of this PR.

**Review by cmoesel — APPROVED** [review]
Two thumbs up. Would recommend to a friend.

File: test/import/FSHImporter.SDRules.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jafeltra) [reply test/import/FSHImporter.SDRules.test.ts:1322]: > On the other hand, when you write the test, you do have to choose a specific kind of SD, right?

Yes, you do. It seems like the majority of the tests in `FSHImporter.SDRules.test.ts` are testing `Profile`, with a handful for `Resource` when testing Resource/Logical Model specific rules.

I was looking at the file history trying to figure out which file came first, and it seemed like SDRules came second, which is why I thought maybe it was trying to ease the number of tests in the Profile file, but that was really just an inference.

> But when we're talking about importing rules, I guess it shouldn't matter if it is a profile, logical, etc... So... ???

Also, when looking at the Importer code, we do specifically have a `visitSdRule` function that is called when importing all four types of StructureDefintions. So maybe to align with the code, the tests in `FSHImporter.SDRules.test.ts` should align with the functionality in `visitSdRule` and any extra functionality, like the rules specific to Logical Models and Resources, are included in separate files for those entities. It feels like maybe that was what we tried to do and things got messy. I'm not sure. And I guess I don't need to be sure for right now.

**Review by mint-thompson — APPROVED** [review]
